parastell/invessel_build.py

```diff
@@ -256,6 +302,150 @@ def generate_components(self):
+                surface.Ribs[-1], next_surface.Ribs[-1], reverse=True
+            )
+            end_cap_end = dagmc.Surface.create(self.dag_model)
+            self.mbc.add_entities(end_cap_end.handle, mb_tris)
+            end_cap_surface_ids.append(
+                list(self.dag_model.surfaces_by_id.keys())[-2:]
+            )
+        self.end_cap_surface_ids = end_cap_surface_ids
+
+    def generate_volumes_pydagmc(self):
+        """Use the curved surface and end cap surface IDs to build the
+        the volumes by applying the correct surface sense to each surface."""
+        # make the volumes and apply surface sense
+        for i in range(len(self.Surfaces) - 1):
+            dagmc.Volume.create(self.dag_model)
+
+        for i, surface_id in enumerate(self.curved_surface_ids):
+            # if it is the first surface, it goes to the implicit complement
+            if i == 0:
+                self.dag_model.surfaces_by_id[surface_id].surf_sense = [
+                    self.dag_model.volumes_by_id[i + 1],
+                    None,
+                ]
+            elif i != len(self.curved_surface_ids) - 1:
+                self.dag_model.surfaces_by_id[surface_id].surf_sense = [
+                    self.dag_model.volumes_by_id[i],
+                    self.dag_model.volumes_by_id[i + 1],
+                ]
+            # if it the last surface it goes to the implicit complement
+            else:
+                self.dag_model.surfaces_by_id[surface_id].surf_sense = [
+                    self.dag_model.volumes_by_id[i],
+                    None,
+                ]
+
+        # all end caps go to the implicit complement.
+        for i, end_cap_ids in enumerate(self.end_cap_surface_ids):
+            for id in end_cap_ids:
+                self.dag_model.surfaces_by_id[id].surf_sense = [
+                    self.dag_model.volumes_by_id[i + 1],
+                    None,
+                ]
+
+    def tag_volumes_with_materials_pydagmc(self):
+        """Tag each volume with the appropriate material name"""
+        for vol, (layer_name, layer_data) in zip(
+            self.dag_model.volumes,
+            list(self.radial_build.radial_build.items())[1:],
+        ):
+
+            mat = layer_data.get("mat_tag", layer_name)
+            group = dagmc.Group.create(self.dag_model, name="mat:" + mat)
+            group.add_set(vol)
+            layer_data["vol_id"] = vol.id
+
+    def generate_components_pydagmc(self):
+        """Use PyDAGMC to build a DAGMC model of the invessel components"""
+        if np.isclose(
+            (self._repeat + 1) * self.radial_build.toroidal_angles[-1], 360
+        ):
+            e = AssertionError(
+                "PyDAGMC workflow does not support modeling of the full 360 "
+                "degrees. Please consider modeling only one period. i.e. set "
+                "'repeat = 0'"
+            )
+            self._logger.error(e.args[0])
+            raise e
```

Comment:
We should probably be considerate of single-period stellarator folks. Although, would it be difficult to support 360-degree modeling (perhaps in a later PR)?
```suggestion
        if np.isclose(
            (self._repeat + 1) * self.radial_build.toroidal_angles[-1], 360
        ):
            e = AssertionError(
                "The PyDAGMC workflow does not support modeling 360-degree"
                "geometries. For configurations with more than one period, please "
                "consider modeling only one period. i.e. set 'repeat = 0'."
            )
            self._logger.error(e.args[0])
            raise e
```

Reply:
I don't think i would be too hard to do, just would need to make sure we handle the matching vertices at the start and end, as well as conditionally not making the end caps. I'd prefer to leave that to a future PR.